devito/ir/equations/equation.py

```diff
@@ -0,0 +1,66 @@
+from collections import OrderedDict
+
+import sympy
+
+from devito.ir.support import Interval, IterationSpace, Stencil
+from devito.symbolics import dimension_sort, indexify
+
+__all__ = ['Eq']
+
+
+class Eq(sympy.Eq):
+
+    """
+    A new SymPy equation with an associated iteration space.
+
+    All :class:`Function` objects within ``expr`` get indexified and thus turned
+    into objects of type :class:`types.Indexed`.
+
+    An iteration space is an object of type :class:`Space`. It represents the
+    data points accessed by the equation along each :class:`Dimension`. The
+    :class:`Dimension`s are extracted directly from the equation.
+    """
+
+    def __new__(cls, input_expr, subs=None):
+        # Sanity check
+        assert isinstance(input_expr, sympy.Eq)
```

Comment:
Should we not also accept our own `Eq` type here? It is highly likely that we might import `Eq` via `from devito import *` in examples and tests.

Reply:
No, at least the way I conceived it. Our own Eq is a lower level concept, performing quite a few low level actions. The ``ir`` modules are not imported from ``devito.__init__``, so it should never be the case that one gets that specific ``Eq`` in scope. Right?

Just to be sure and to make things clearer, I could change the name, such as ``LoweredEq`` or something along these lines.

Reply:
Yes, so a name change might be best then, as this is in no way obvious atm. My main worry was what happens if we do `LoweredEq(input_expr=LoweredEq(some expression))`, which we need to assert against (I take it it `isinstance(LoweredEq, Eq)`, right?

Reply:
I'll change the name.
The assert would still work; however, there's no point in... "lowering an already lowered expression"
Said that, I've already extended this particular piece of code in my current local branch

Reply:
> there's no point in... "lowering an already lowered expression"

Exactly, and the assert won't catch such foolishness either. Hence my comment about both classes having to be very distinct, and not sharing a name. Ideally we would even catch if someone attempted this, but that's not strictly necessary, I guess.